doc/introduction/measurements.rst

```diff
@@ -125,6 +124,50 @@ accept observables as arguments,
 including :func:`~.pennylane.expval`, :func:`~.pennylane.var`,
 and :func:`~.pennylane.sample`.
 
+Counts
+------
+
+To avoid dealing with long arrays for the larger numbers of shots, one can pass an argument counts=True
+to :func:`~pennylane.sample`. In this case the result will be a dictionnary indicating numbers of occurences for each
+unique sample. The previous example will be modified as follows:
+
+.. code-block:: python
+
+    dev = qml.device("default.qubit", wires=2, shots=1000)
+
+    @qml.qnode(dev)
+    def circuit():
+        qml.Hadamard(wires=0)
+        qml.CNOT(wires=[0, 1])
+        # passing the counts flag
+        return qml.sample(qml.PauliZ(0), counts=True), qml.sample(qml.PauliZ(1), counts=True)
+
+After executing the circuit we can directly see how often each value was measured:
+        
+>>> result = circuit()
+>>> print(result)
+[tensor({-1: 526, 1: 474}, dtype=object, requires_grad=True)
```

Comment:
Any thoughts on why the data structure is different here? I.e., the result is a list of tensors created from dictionaries instead of a single dictionary?

Reply:
It's not seen from the examples (because of how print() interacts with tensors), but in all the cases the function returns a tensor, that contains either a dict or other tensors of dictionaries. I went with this option in order to preserve the consistency with the return object of the counts=False case, but I wasn't sure if it's the best choice.
So, you'd suggest to avoid the tensors and return a dictionary/list of dictionaries?

Reply:
Oh, that makes sense, thanks for clarifying that. :)

> So, you'd suggest to avoid the tensors and return a dictionary/list of dictionaries?

I think that would make sense, though feel free to argue against it. :slightly_smiling_face: Usually, the reason why we cast results into PennyLane NumPy arrays (tensors) is to provide differentiability for the outputs of QNodes. `qml.sample` is an exception here, it is just convenient to have `qml.sample`'s output with `counts=False` to be arraylike. When `counts=True` is set, however, perhaps a simple dictionary/list of dictionaries would make more sense. One thing to check here though might be how such a solution would work with composite measurements, e.g., would having
```py
return qml.probs(wires=[0,1]), qml.sample(wires=2, counts=True)
```
in a quantum function lead to casting issues?

Reply:
Indeed there are three cases:

1. Only counts. I agree, the most handy option here would be to return a dict.
2. Combination of counts with counts. I agree again, let's return a list of dicts.
3. Counts + something else. Something else may need gradient in some following computations, let's return tensor of tensors. In the way it's done now, the counts dict will be in a tensor as well, but it's easy to extract it, so why not.

The only downside is that it's cumbersome to have three different return types.

Another way would be to pass an option what to return, but it seems too much, including two variables just for a regrouping of counts (at this moment it may be simpler for a user to import and use Counter from collections).

Reply:
I chose the way 3, it's implemented in several latest commits